core/controllers/reader_test.py

```diff
@@ -362,3 +363,169 @@ def test_ratings_by_different_users(self):
             ratings['overall_ratings'],
             {'1': 0, '2': 0, '3': 0, '4': 2, '5': 0})
         self.logout()
+
+
+class FlagExplorationHandlerTests(test_utils.GenericTestBase):
+
+    EXP_ID = '0'
+
+    def setUp(self):
+        super(FlagExplorationHandlerTests, self).setUp()
+
+        # Register users.
+        self.signup(self.MODERATOR_EMAIL, self.MODERATOR_USERNAME)
+        self.signup(self.EDITOR_EMAIL, self.EDITOR_USERNAME)
+        self.signup(self.NEW_USER_EMAIL, self.NEW_USER_USERNAME)
+
+        self.editor_id = self.get_user_id_from_email(self.EDITOR_EMAIL)
+        self.new_user_id = self.get_user_id_from_email(self.NEW_USER_EMAIL)
+        self.moderator_id = self.get_user_id_from_email(self.MODERATOR_EMAIL)
+        self.set_moderators([self.MODERATOR_USERNAME])
+
+        # Load exploration 0.
+        exp_services.delete_demo(self.EXP_ID)
+        exp_services.load_demo(self.EXP_ID)
+
+        # Login and create exploration.
+        self.login(self.EDITOR_EMAIL)
+
+        # Create exploration.
+        self.save_new_valid_exploration(
+            self.EXP_ID, self.editor_id,
+            title='Welcome to Oppia!',
+            category='This is just a spam category',
+            objective='Test a spam exploration.')
+        self.can_send_emails_ctx = self.swap(
+            feconf, 'CAN_SEND_EMAILS', True)
+        exploration = exp_services.get_exploration_by_id(self.EXP_ID)
+        init_state = exploration.states[exploration.init_state_name]
+        init_interaction = init_state.interaction
+        init_interaction.default_outcome.dest = exploration.init_state_name
+        exploration.add_states(['State A', 'State 2', 'State 3'])
+        exploration.states['State A'].update_interaction_id('TextInput')
+        exploration.states['State 2'].update_interaction_id('TextInput')
+        exploration.states['State 3'].update_interaction_id('TextInput')
+        exp_services._save_exploration(self.editor_id, exploration, '', [])  # pylint: disable=protected-access
+        rights_manager.publish_exploration(self.editor_id, self.EXP_ID)
+        rights_manager.assign_role_for_exploration(
+            self.editor_id, self.EXP_ID, self.moderator_id,
+            rights_manager.ROLE_EDITOR)
+        self.logout()
+
+        # Login and flag exploration.
```

Comment:
I was able to do the test when user is logged in but when user is not logged in, I get some errors. I have written tests in moderator_services if the user_id is none.